include/ts/apidefs.h.in

```diff
@@ -1070,6 +1070,11 @@ typedef struct TSHttpConnectOptions_s {
   TSIOBufferWaterMark buffer_water_mark;
 } TSHttpConnectOptions;
 
+typedef struct TSDbgCtl_s {
+  char volatile on; // Flag
```

Comment:
The reality is that ATS often relies on de facto portability, rather than portability guaranteed by the Standard.

I guess memory ordering has been added to straight C, but I don't know the specifics, I'll have to research this.  The point of the volatile is to emphasize that that the on flag should not be register cached.  But it seems implausible the compiler could do that indefinitely anyway.

Reply:
It's correct that (when using Debug() instead of Dbg() ) there's a significant increase in loaded data segment size,  But, it's a delta on top of the high resource use of trace output even with the old design.  Which seems worthwhile, to make tracing usable in production.

Reply:
I agree this is a worthwhile trade off for performant debugging! It's good stuff.

I also think that in this case, any changes to `on` values will be seen "soon enough" by other threads and that volatile just isn't necessary here(especially since its accessed through a non-volatile pointer).

Reply:
Although it's unlikely, this code illustrates the sort of optimization that could happen to a Debug statement in a loop, if we don't do something to prevent the compiler from register-caching the 'on' flag:  https://godbolt.org/z/r4K9TP1qW

Reply:
The alternate to simply relying on volatile would be to use atomic_load_explicit() with relaxed memory order.  But that would mean we'd have to require plugins to be build with C11 support, which I don't think we've done up till now.

Reply:
Note that the Standard says that the behavior of reinterpret_cast is undefined, but ATS uses it over 1,400 times.

Reply:
This is closer to (but still not quite) the situation we have [here](https://godbolt.org/z/EEYP3YMdj).  We are accessing a volatile through a (non-volatile) static pointer.  The compiler must read from memory at least one time when the containing function is called.  For the non-volatile case, the compiler can then assume the value won't change until another function is called or the original function is called again.  This is what I meant by it will be re-read "soon enough".

All that said, if you are happier with the volatile qualifier and no-one else objects, I'm not overly concerned about it.